README
WWW-ConfixxBackup version 0.03
==============================

This module aims to simplify making backups of Websites using Confixx.
Sometimes it is forgotten to make backups.

This module can handle either the same login data for Confixx and FTP or
different login data.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  WWW::Mechanize
  Net::FTP

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2006 by Renee Baecker

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension WWW::ConfixxBackup.

0.01  Wed Feb 22 11:07:34 2006
	- original version; created by h2xs 1.23 with options
		-b 5.6.1 -XA -n WWW::ConfixxBackup
        
0.02  Sat Jun 17 23:58:52 2006
    - bug fixed in WWW::ConfixxBackup::Confixx.pm
    - new method wait()
    
0.03  Tue Jun 20 20:04:33 2006
    - bug in test script fixed

0.05  Fri Dec 08 10:17:16 2006
    - added: 97_manifest.t
    - added: 98_boilerplate.t
    - added: 99_pod.t
    - added: 99_pod_coverage.t
    - added: LICENSE field in Makefile.PL

0.06  Mon Dec 18 11:00:05 2006
    - bugfix: 99_pod_coverage.t

0.07  Tue Sep 18 11:11:11 2007
    - bugfix: ConfixxBackup.pm / now you can pass parameters to constructor
    - pod: fixed some typos in documentation

0.08  Fri Dec 28 09:12:13 2007
    + added new functionality:
       + confixx_version
       + available_confixx_versions
       + default_confixx_version
    + added new test:
       + 08_confixx_version.t

0.0801 Sat Dec 29 12:55:31 2007
    + use binary mode for FTP transfer

0.09  Wed Jan 02 13:36:13 2008
    + added "file_prefix" to enable parallel downloads
    + added "_detect_version" for automatic "confixx_version" detection (when confixx_version is not set)
    + added tests:
        * t/09_file_prefix.t
        * t/10_check_can_ok.t
        * t/11_can_ok_Confixx.t
        * t/12_can_ok_FTP.t
        * t/13_detect_version.t
        * t/14_detect_version_2.t
    + added debug() to WWW::ConfixxBackup, WWW::ConfixxBackup::Confixx, WWW::ConfixxBackup::FTP
    + added DEBUG() to WWW::ConfixxBackup, WWW::ConfixxBackup::Confixx, WWW::ConfixxBackup::FTP
    + added some debugging stuff

0.10  Mon Nov 03 11:57:23
    + fixed a bug in 06_*.t that caused the test to fail


Makefile.PL
use 5.006001;
use ExtUtils::MakeMaker;

for $i(0..scalar(@ARGV)-1){
  if($ARGV[$i] =~ /^--userfile$/){
    if(open(my $fh,">",'./t/userfile.txt')){
      if(open(my $read,"<",$ARGV[$i+1])){
        while(my $line = <$read>){
          print $fh $line;
        }
        close $read;
      }
      close $fh;
    }
    else{
      warn "Can't write userfile";
    }
  }
}

WriteMakefile(
    NAME              => 'WWW::ConfixxBackup',
    VERSION_FROM      => 'lib/WWW/ConfixxBackup.pm', # finds $VERSION
    PREREQ_PM         => {
                           WWW::Mechanize      => 0,
                           Net::FTP            => 0,
                           HTTP::Cookies       => 0,
                           Test::CheckManifest => 1.0,
                           Test::More          => 0,
                           FindBin             => 0,
                         }, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/WWW/ConfixxBackup.pm', # retrieve abstract from module
       AUTHOR         => 'Renee Baecker <module@renee-baecker.de>') : ()),
    ($ExtUtils::MakeMaker::VERSION >= 6.31 ?
      (LICENSE        => 'perl') : ()),
);


lib/WWW/ConfixxBackup/FTP.pm
package WWW::ConfixxBackup::FTP;

use strict;
use warnings;
use Net::FTP;

our $VERSION = '0.01';

sub new{
  my ($class,%args) = @_;
  my $self = {};
  bless $self,$class;
  
  $self->password($args{password});
  $self->user($args{user});
  $self->server($args{server});
  
  return $self;
}# new

sub login{
  my ($self) = @_;
  if(ref($self->ftp) eq 'Net::FTP'){
    $self->ftp()->login($self->user,$self->password) or return 0;
  }
  else{
    return 0;
  }
  return 1;
}# login

sub download{
  my ($self,$path) = @_;
  $path ||= '.';
  eval{
    $self->ftp()->cwd('/backup');
    for(qw/mysql.tar.gz files.tar.gz html.tar.gz/){
      $self->ftp()->get($_,$path.'/'.$_);
    }
  };
  return undef if($@);
  return 1;
}# download

sub password{
  my ($self,$pwd) = @_;
  $self->{password} = $pwd if(defined $pwd);
  return $self->{password};
}# password

sub user{
  my ($self,$user) = @_;
  $self->{user} = $user if(defined $user);
  return $self->{user};
}# user

sub server{
  my ($self,$server) = @_;
  $self->{server} = $server if(defined $server);
  return $self->{server};
}# server

sub ftp{
  my ($self) = @_;
  unless(ref($self->{ftp}) eq 'Net::FTP'){
    $self->{ftp} = Net::FTP->new($self->server);
  }
  return $self->{ftp};
}# mech

sub DESTROY{
  my ($self) = @_;
  $self->ftp()->quit() if(ref($self->ftp) eq 'Net::FTP');
}

1;

__END__
=pod

=head1 NAME

WWW::ConfixxBackup::FTP - the FTP client for WWW::ConfixxBackup

=head1 SYNOPSIS

  use WWW::ConfixxBackup::FTP;
  
  my $ftp = WWW::ConfixxBackup::FTP->new(server => 'server');
  $ftp->user('username');
  $ftp->password('password');
  
  $ftp->login;
  $ftp->download('/path/for/download');

=head1 METHODS

=head2 new

=head2 user

=head2 password

=head2 server

=head2 login

=head2 download

=head2 ftp

=head1 SEE ALSO

  Net::FTP
  
=head1 AUTHOR

Renee Baecker, E<lt>module@renee-baecker.deE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 by Renee Baecker

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut

lib/WWW/ConfixxBackup/Confixx.pm
package WWW::ConfixxBackup::Confixx;

use strict;
use warnings;
use WWW::Mechanize;
use HTTP::Cookies;
use HTTP::Request;

our $VERSION = '0.03';

my %map = (
  'confixx2.0' =>
    {
      backup_id    => '',
      backup_html  => '',
      backup_files => '',
      backup_mysql => '',
      html         =>  1,
      files        =>  1,
      mysql        =>  1,
    },
  'confixx3.0' =>
    {
      'selectAll' => 1,
      'backup[]'  => ['html','files','mysql'],
      action      => 'backup',
      destination => '/backup',
    }
);
my $default_version = 'confixx3.0';

sub new{
  my ($class,%args) = @_;
  my $self = {};
  bless $self,$class;
  
  $self->password($args{password});
  $self->user($args{user});
  $self->server($args{server});
  $self->confixx_version( $args{confixx_version} || $default_version );
  
  return $self;
}# new

sub login{
  my ($self) = @_;
  
  if(ref($self->mech) eq 'WWW::Mechanize'){
    $self->mech->post($self->server . '/login.php',
                                 {username => $self->user, 
                                  password => $self->password,});
    $self->mech->get($self->server . '/user/' . $self->user . '/');
        
    return 0 unless($self->mech->success);
  }
  return 1;
}# login

sub _detect_version{
    my ($self) = @_;
    
    $self->mech->get( $self->server . '/user/' . $self->user . '/tools_backup.php' );
    warn $self->mech->content;
}

sub backup{
  my ($self) = @_;
  $self->mech->post($self->server . '/user/' . $self->user . '/tools_backup2.php',
        $map{$self->confixx_version},
  );
  
  return 0 unless($self->mech->success);
  return 1;
}# create_backup

sub confixx_version{
    my ($self,$version) = @_;
    
    $self->{__version} = $version if defined $version;
    return $self->{__version};
}

sub available_confixx_versions{
    return sort keys %map;
}

sub password{
  my ($self,$pwd) = @_;
  $self->{password} = $pwd if(defined $pwd);
  return $self->{password};
}# password

sub user{
  my ($self,$user) = @_;
  $self->{user} = $user if(defined $user);
  return $self->{user};
}# user

sub server{
  my ($self,$server) = @_;
  $self->{server} = $server if(defined $server);
  return $self->{server};
}# server

sub mech{
  my ($self) = @_;
  unless(ref($self->{mechanizer}) eq 'WWW::Mechanize'){
    $self->{cookie_jar} = HTTP::Cookies->new();
    $self->{mechanizer} = WWW::Mechanize->new(
                                    quiet       => 1,
                                    onwarn      => \&mech_warnings,
                                    stack_depth => 1,
                                    cookie_jar  => $self->{cookie_jar},);
    $self->{mechanizer}->get($self->server);
    
    return [] unless($self->{mechanizer}->success);
  }
  return $self->{mechanizer};
}# mech

sub default_version{
    return $default_version;
}

sub mech_warnings{
  #print STDERR "HALLO";
}# mech_warnings

1;

__END__

=pod

=head1 NAME

WWW::ConfixxBackup::Confixx - the Confixx mechanism for WWW::ConfixxBackup

=head1 SYNOPSIS

=head1 METHODS

=head2 new

=head2 user

=head2 password

=head2 server

=head2 backup

=head2 mech_warnings

=head2 mech

=head2 login

=head2 confixx_version

=over 4

=item * confixx2.0

=over 4

=item * backup_id

=item * backup_html

=item * backup_files

=item * backup_mysql

=item * html

=item * files

=item * mysql

=back

=item * confixx3.0

=over 4

=item * selectAll

=item * backup[]

=over 4

=item * html

=item * files

=item * mysql

=back

=item * destination

=item * action

=back

=back

=head2 available_confixx_versions

returns a list of all confixx versions (to be precisely versions of tools_backup2.php) 
that are supported by WWW::ConfixxBackup

=head2 default_version

returns the default value for confixx_version

=head1 SEE ALSO

  Net::FTP
  
=head1 AUTHOR

Renee Baecker, E<lt>module@renee-baecker.deE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 by Renee Baecker

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/WWW/ConfixxBackup.pm
package WWW::ConfixxBackup;

use 5.006001;
use strict;
use warnings;
use WWW::ConfixxBackup::Confixx;
use WWW::ConfixxBackup::FTP;

our $VERSION = '0.10';

sub new{
    my ($class,%args) = @_;
    my $self = bless {},$class;
    
    $self->waiter(120);
    $self->file_prefix( '' );
    
    $self->debug( sub{ print join "\n", @_ } );
    $self->DEBUG(0);
    
    my @unwanted = (qr/login$/, qr/^backup/, qr/download$/, qr/^_/, qr/errstr/);
    
    for my $key ( keys %args ){
        next if grep{ $key =~ $_ }@unwanted;
        if( my $sub = $self->can( $key ) ){
            $sub->( $self, $args{$key} );
        }
    }
    
    return $self;
}# new

sub http_proxy{
    my ($self,$proxy) = @_;
    
    $self->{_http_proxy} = $proxy if @_ == 2;
    return $self->{_http_proxy};
}

sub ftp_user{
    my ($self,$user) = @_;
    $self->{ftp_user} = $user if(defined $user);
    return $self->{ftp_user};
}# ftp_user

sub ftp_server{
    my ($self,$server) = @_;
    $self->{ftp_server} = $server if(defined $server);
    return $self->{ftp_server};
}# ftp_server

sub ftp_password{
    my ($self,$password) = @_;
    $self->{ftp_password} = $password if(defined $password);
    return $self->{ftp_password};
}#ftp_password

sub user{
    my ($self,$user) = @_;
    if(defined $user){
       $self->ftp_user($user);
       $self->confixx_user($user);
    }
    return $self->ftp_user;
}# user

sub server{
    my ($self,$server) = @_;
    if(defined $server){
        $self->ftp_server($server);
        $self->confixx_server($server);
    }
    return $self->ftp_server;
}# server

sub password{
    my ($self,$password) = @_;
    if(defined $password){
        $self->ftp_password($password);
        $self->confixx_password($password);
    }
    return $self->ftp_password;
}# password

sub confixx_user{
    my ($self,$user) = @_;
    $self->{confixx_user} = $user if(defined $user);
    return $self->{confixx_user};
}# confixx_user

sub confixx_server{
    my ($self,$server) = @_;
    $self->{confixx_server} = $server if(defined $server);
    return $self->{confixx_server};
}# confixx_server

sub confixx_password{
    my ($self,$password) = @_;
    $self->{confixx_password} = $password if(defined $password);
    return $self->{confixx_password};
}# confixx_pasword

sub confixx_version{
    my ($self,$version) = @_;
    $self->{_confixx_version} = $version if defined $version;
    return $self->{_confixx_version};
}

sub available_confixx_versions{
    my ($self) = @_;
    return WWW::ConfixxBackup::Confixx->available_confixx_versions;
}

sub default_confixx_version{
    return WWW::ConfixxBackup::Confixx->default_version;
}

sub login{
    my ($self) = @_;
    $self->_reset_errstr;
    $self->ftp_login     or $self->_add_errstr('' . $self->ftp_server);
    $self->confixx_login or $self->_add_errstr('' . $self->confixx_server);
    
    unless($self->errstr){
        return 1;
    }
    
    return;
}# login

sub ftp_login{
    my ($self) = @_;
    $self->{FTP} = WWW::ConfixxBackup::FTP->new(
        user     => $self->ftp_user,
        password => $self->ftp_password,
        server   => $self->ftp_server,
    );
    $self->{FTP}->debug( $self->debug );
    $self->{FTP}->DEBUG( $self->DEBUG );
    $self->{FTP}->login;
    
    return if ref($self->{FTP}->ftp) ne 'Net::FTP';
    return 1;
}# ftp_login

sub confixx_login{
    my ($self) = @_;
    $self->{CONFIXX} = WWW::ConfixxBackup::Confixx->new(
        user     => $self->confixx_user,
        password => $self->confixx_password,
        server   => $self->confixx_server,
    );
    $self->{CONFIXX}->debug( $self->debug );
    $self->{CONFIXX}->DEBUG( $self->DEBUG );
    $self->{CONFIXX}->proxy( $self->http_proxy ) if $self->http_proxy;
    $self->{CONFIXX}->login;
    
    if( ref($self->{CONFIXX}->mech) ne 'WWW::Mechanize' ){
        return;
    }
    return 1;
}# confixx_login

sub file_prefix{
    my ($self,$prefix) = @_;
    
    $self->{__prefix} = $prefix if scalar @_ > 1;
    return $self->{__prefix};
}

sub backup_download{
    my ($self,$path) = @_;
    
    $path ||= '.';
    $self->_reset_errstr();
    
    unless($self->{CONFIXX}){
        $self->confixx_login or $self->_add_errstr('Can\'t login to Confixx');
    }
    
    unless($self->{FTP}){
        $self->ftp_login or $self->_add_errstr('Can\'t login to FTP server');
    }
    
    $self->{CONFIXX}->confixx_version( $self->confixx_version );
    if(defined $path && $self->{CONFIXX} && $self->{FTP}){
        $self->{CONFIXX}->backup() or $self->_add_errstr('Can\'t create backup');
        sleep($self->{WAIT});
        $self->{FTP}->prefix( $self->file_prefix );
        $self->{FTP}->download($path) or $self->_add_errstr('Can\'t download');
    }
    
    unless($self->errstr){
        return 1;
    }
    
    return 0;
}# backup_download

sub _reset_errstr{
    my ($self) = @_;
    $self->{errstr} = '';
}

sub _add_errstr{
    my ($self,$msg) = @_;
    $self->{errstr} .= $msg if(defined $msg);
}

sub errstr{
    my ($self) = @_;
    return $self->{errstr}."\n";
}

sub waiter{
    my ($self,$wait) = @_;
    $self->{WAIT} = $wait if(defined $wait);
    return $self->{WAIT};
}# wait

sub detect_version{
    my ($self) = @_;
    
    $self->_reset_errstr;
    
    unless($self->{CONFIXX}){
        $self->confixx_login or
                $self->add_errstr('Can\'t login to Confixx' . $self->confixx_server);
    }
    if($self->{CONFIXX}){
        $self->{CONFIXX}->_detect_version;
        my $version = $self->{CONFIXX}->confixx_version;
        $self->confixx_version( $version );
        return $version;
    }

    return;
}

sub debug{
    my ($self,$coderef) = @_;
    
    if( @_ == 2 ){
        $self->DEBUG(1);
        $self->{__debug} = $coderef;
    }

    return $self->{__debug};
}

sub DEBUG{
    my ($self,$bool) = @_;
    
    if( @_ == 2 ){
        $self->{__DEBUG} = $bool;
        $self->{CONFIXX}->DEBUG( $bool ) if $self->{CONFIXX};
        $self->{FTP}->DEBUG( $bool )     if $self->{FTP};
    }

    return $self->{__DEBUG};
}

sub download{
    my ($self,$path) = @_;
    
    $self->_reset_errstr;
    
    unless($self->{FTP}){
      $self->ftp_login or 
              $self->add_errstr('Can\'t login to FTP server' . $self->ftp_server);
    }
    if($self->{FTP}){
      $self->{FTP}->download($path) or
              $self->add_errst('Can\'t download backup files');
    }
    
    unless($self->errstr){
        return 1;
    }
    return;
}# download

sub backup{
    my ($self) = @_;
    
    $self->_reset_errstr;
    
    unless($self->{CONFIXX}){
        $self->confixx_login or
                $self->add_errstr('Can\'t login to Confixx' . $self->confixx_server);
    }
    if($self->{CONFIXX}){
        $self->{CONFIXX}->confixx_version( $self->confixx_version );
        $self->{CONFIXX}->backup or
                $self->add_errstr('Can\'t create backup');
    }
    
    unless($self->errstr){
        return 1;
    }
    return 0;
}# backup


# Preloaded methods go here.

1;

__END__

=head1 NAME

WWW::ConfixxBackup - Create Backups with Confixx and download them via FTP

=head1 SYNOPSIS

  use WWW::ConfixxBackup;
  
  #shortes way (and Confixx and FTP use the same login data)
  my $backup = WWW::ConfixxBackup->new(user => 'user', password => 'user', server => 'server');
  my $path = './backups/today/';
  $backup->backup_download($path);
  
  #longer way (and different Confixx and FTP login data)
  my $backup = WWW::ConfixxBackup->new();
  $backup->ftp_user('ftp_user');
  $backup->ftp_password('ftp_password');
  $backup->ftp_server('server');
  $backup->ftp_login();
  $backup->confixx_user('confixx_user');
  $backup->confixx_password('confixx_password');
  $backup->confixx_server('confixx_server');
  $backup->confixx_login();
  $backup->confixx_version( 'confixx3.0' );
  $backup->backup();
  $backup->download($path);
  $backup->waiter($seconds);

=head1 DESCRIPTION

This module aims to simplify backups via Confixx and FTP. It logs in Confixx,
creates the backups and downloads the backups via FTP.

=head2 METHODS

=head3 new

  my $backup = WWW::ConfixxBackup->new();
  
creates a new C<WWW::ConfixxBackup> object.

=head3 user

  $backup->user('username');
  print $backup->user();

=head3 password

  $backup->password('password');
  print $backup->password();

=head3 server

  $backup->server('server');
  print $backup->server();

=head3 confixx_user

  $backup->confixx_user('confixx_username');
  print $backup->confixx_user();

=head3 confixx_password

  $backup->confixx_password('confixx_password');
  print $backup->confixx_password();

=head3 confixx_server

  $backup->confixx_server('confixx_server');
  print $backup->confixx_server();

=head3 confixx_version

  $backup->confixx_version( 'confixx3.0' );
  print $backup->confixx_version;

The parameters for the confixx script have been changed. If you know which version
is used, you can set the versionstring. If you don't know it, you ask for it via
C<detect_version>.

=head3 detect_version

  print $backup->detect_version;

=head3 available_confixx_versions

returns a list of all confixx versions (to be precisely versions of tools_backup2.php) 
that are supported by WWW::ConfixxBackup

=head3 default_confixx_version

returns the default value for confixx_version

=head3 http_proxy

  $backup->http_proxy( $proxy );

sets a proxy for HTTP requests

=head3 ftp_user

  $backup->ftp_user('ftp_user');
  print $backup->ftp_user();

=head3 ftp_password

  $backup->ftp_password('ftp_password');
  print $backup->ftp_password();

=head3 ftp_server

  $backup->ftp_server('ftp_server');
  print $backup->ftp_server();

=head3 file_prefix

  $backup->file_prefix( 'account_name' );
  print $backup->file_prefix

set/get a prefix for the backup files. This is necessary if you do parallel
backups.

=head3 confixx_login

  $backup->confixx_login();

=head3 ftp_login

  $backup->ftp_login();

login on FTP server

=head3 login

login on Confixx server and FTP server

=head3 backup

  $backup->backup();

Logs in to Confixx and creates the backups

=head3 download

  $backup->download('/path/to/directory');

downloads the three files that are created by Confixx:

=over 4

=item * mysql.tar.gz

=item * html.tar.gz

=item * files.tar.gz

=back

to the given path. If path is omitted, the files are downloaded to the
current directory.

=head3 backup_download

  $backup->backup_download('/path/to/directory/');

logs in to Confixx, create the backup files and downloads the three files that 
are created by Confixx:

=over 4

=item * mysql.tar.gz

=item * html.tar.gz

=item * files.tar.gz

=back

to the given path. If path is omitted, the files are downloaded to the
current directory.

=head3 waiter

  $backup->waiter(100);

sets the value for the sleep-time in seconds

=head3 errstr

  print $backup->errstr();

returns an error message when an error occured

=head3 debug

If you want more debugging, you can use your own subroutine. This subroutine will
get the debug message as an argument.

  $backup->debug( sub{ print @_ } );

=head3 DEBUG

With C<DEBUG> you can switch on and off the debugging mode. If you don't use your
own subroutine (see L<debug>), a default subroutine is used that just prints the
messages to STDOUT

  $backup->DEBUG(1);

The debugging mode is turned off by default.

=head1 SEE ALSO

  WWW::ConfixxBackup::Confixx
  WWW::ConfixxBackup::FTP
  WWW::Mechanize
  Net::FTP

=head1 AUTHOR

Renee Baecker, E<lt>module@renee-baecker.deE<gt>

=head1 COPYRIGHT AND LICENSE

2006 - 2008 by Renee Baecker

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.


=cut
